src/MongrationDotNet/ClientSideDocumentMigration.cs

```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using MongoDB.Bson;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Driver;
+
+namespace MongrationDotNet
+{
+    public abstract class ClientSideDocumentMigration : Migration
+    {
+        public override string Type { get; } = Constants.ClientSideDocumentMigrationType;
+        public abstract string CollectionName { get; }
+        public virtual FilterDefinition<BsonDocument> SearchFilters { get; set; } = FilterDefinition<BsonDocument>.Empty;
+        public virtual string UpdateFilterField { get; } = "_id";
+
+        public override async Task ExecuteAsync(IMongoDatabase database, ILogger logger)
+        {
+            logger?.LogInformation(LoggingEvents.DocumentMigrationStarted, "Migration started for {collection}",
+                CollectionName);
+
+            var collection = database.GetCollection<BsonDocument>(CollectionName);
+
+            var documents = await collection.Find(SearchFilters).ToListAsync();
+            foreach (var document in documents)
+            {
+                var migratedDocument = MigrateDocument(document);
+                document.AsBsonDocument.TryGetElement(UpdateFilterField, out var bsonValue);
+                await collection.ReplaceOneAsync(new BsonDocument(UpdateFilterField, bsonValue.Value), migratedDocument,
+                    new ReplaceOptions { IsUpsert = true });
+            }
```

Comment:
Found another race condition i think. Say you have a million docs to migrate client side. while the migration is running the live server inserts a a new document into that collection in the old data format. This doc will not be migrated. Insetad we should loop untill there are no more documents remaining. Also we should watch out for loading the entire db into memory in one go using batches.

```
//props
public virtual int BatchSize { get; } = int.MaxValue;


//in ExecuteAsync
do {
            int updated = 0;
            var documents = await collection.Find(SearchFilters).Take(batchSize).ToListAsync();
            foreach (var document in documents)
            {
                var migratedDocument = MigrateDocument(document);
                document.AsBsonDocument.TryGetElement(UpdateFilterField, out var bsonValue);
                await collection.ReplaceOneAsync(new BsonDocument(UpdateFilterField, bsonValue.Value), migratedDocument,
                    new ReplaceOptions { IsUpsert = true });
              updated++;
            }
} while (updated > 0)
```

Reply:
This is creating an infinite loop. I have tweaked the loop a little. I think we are still meeting the requirements.

Reply:
Will only inifite loop if you give a bad query, I was thinking a query like "give me all docs that have property named 'oldProp'" and then the migration would remove 'oldProb' so that when done there would be nothing left in the collection. But i accept there may be uses where you just want to run it on the entire collection once.